dali/pipeline/operator/sequence_operator.h

```diff
@@ -0,0 +1,606 @@
+      auto layout_prefix = expand_desc.ExpandedLayout();
+      if (ws.template OutputIsType<GPUBackend>(output_idx)) {
+        SetOutputLayout<GPUBackend>(ws, output_idx, layout_prefix);
+      } else {
+        SetOutputLayout<CPUBackend>(ws, output_idx, layout_prefix);
+      }
+    }
+  }
+
+  /**
+   * @brief Coalesces batch of frames shapes into batch of sequence shapes.
+   *
+   * For example if in the input there is a ``FCHW`` batch of shape
+   * ``[{1, 3, 20, 20}, {2, 2, 40, 20}]`` with ``FC`` expandable layout, the inferred output shape
+   * (if any) is expected to be of the form
+   * ``[shape_1, shape_1, shape_1, shape_2, shape_2, shape_2, shape_2]``, so that it can be
+   * coalesced into ``[{1, 3, shape_1...}, {2, 2, shape_2...}]``.
+   */
+  virtual void CoalesceOutputShapes(std::vector<OutputDesc> &output_desc,
+                                    const workspace_t<Backend> &ws) {
+    for (size_t output_idx = 0; output_idx < output_desc.size(); output_idx++) {
+      const auto &expand_desc = GetOutputExpandDesc(ws, output_idx);
+      const auto &shape = output_desc[output_idx].shape;
+      DALI_ENFORCE(static_cast<size_t>(shape.num_samples()) == expand_desc.NumExpanded(),
+                   make_string("Unexpected number of frames inferred in the operator for output ",
+                               output_idx, ". Expected ", expand_desc.NumExpanded(),
+                               " but the operator returned ", shape.num_samples(), "."));
+      output_desc[output_idx].shape =
+          sequence_utils::fold_outermost_like(shape, expand_desc.DimsToExpand());
+    }
+  }
+
+  const ExpandDesc &GetInputExpandDesc(int input_idx) const {
+    DALI_ENFORCE(0 <= input_idx && static_cast<size_t>(input_idx) < input_expand_desc_.size());
+    return input_expand_desc_[input_idx];
+  }
+
+  void SetupSequenceOperator(const workspace_t<Backend> &ws) {
+    auto num_inputs = ws.NumInput();
+    input_expand_desc_.clear();
+    input_expand_desc_.reserve(num_inputs);
+    for (int input_idx = 0; input_idx < num_inputs; input_idx++) {
+      const auto &input_shape = ws.GetInputShape(input_idx);
+      const auto &layout = GetInputLayout(ws, input_idx);
+      input_expand_desc_.emplace_back(input_shape, layout, ShouldExpandChannels(input_idx));
+    }
+    expand_like_idx_ = InferExpandableInputIdx(ws);
+  }
+
+  int InferExpandableInputIdx(const workspace_t<Backend> &ws) {
+    auto num_inputs = ws.NumInput();
+    for (int input_idx = 0; input_idx < num_inputs; input_idx++) {
+      if (input_expand_desc_[input_idx].NumDimsToExpand() > 0) {
+        return input_idx;
+      }
+    }
+    return -1;
+  }
+
+  /**
+   * @brief Returns the index of the operator input which is expandable and other
+   * inputs should be expanded or broadcasted in a manner consistent with the
+   * given input. If no input is expandable, returns -1.
+   */
+  int GetReferentialInputIdx() const {
```

Comment:
```suggestion
  int GetReferenceInputIdx() const {
```
ReferentialInputIdx reads like "index of referential input" and a "referential input" would be an input that contain references (or is in some other way related to references).

Reply:
done